ts/types.go

```diff
@@ -45,27 +50,70 @@ type id struct {
 }
 
 // Data returns the binary representation of an ID
-func (id *id) Data() []byte {
-	return id.data
+func (v *id) Data() []byte {
+	return v.data
 }
 
-func (id *id) Equal(value ID) bool {
-	return bytes.Equal(id.Data(), value.Data())
+func (v *id) Equal(value ID) bool {
+	return bytes.Equal(v.Data(), value.Data())
 }
 
 var null = Hash{}
 
 // Hash returns (and calculates, if needed) the hash for an ID
-func (id *id) Hash() Hash {
-	if bytes.Equal(id.hash[:], null[:]) {
-		id.hash = md5.Sum(id.data)
+func (v *id) Hash() Hash {
+	if bytes.Equal(v.hash[:], null[:]) {
+		v.hash = md5.Sum(v.data)
 	}
 
-	return Hash(id.hash)
+	return Hash(v.hash)
+}
+
+func (v *id) String() string {
+	return string(v.data)
+}
+
+// Clone constructs a new ID identical to the current
+func (v *id) Clone() ID {
+	return &id{data: v.data, hash: v.hash}
+}
+
+// Pool represents an object pool of IDs
+type Pool struct {
+	pool pool.ObjectPool
+}
+
+// NewPool constructs a new Pool
+func NewPool(options pool.ObjectPoolOptions) *Pool {
+	p := pool.NewObjectPool(options)
+	p.Init(func() interface{} { return &id{} })
+
+	return &Pool{pool: p}
+}
+
+// BinaryID returns a new ID based on a binary blob
+func (p *Pool) BinaryID(v []byte) ID {
+	id := p.pool.Get().(*id)
+	id.data = v
+
+	return id
 }
 
-func (id *id) String() string {
-	return string(id.data)
+// StringID returns a new ID based on a string
+func (p *Pool) StringID(v string) ID {
+	id := p.pool.Get().(*id)
+	id.data = []byte(v)
+
+	return id
+}
+
+// Release returns a set of IDs back into the pool
+func (p *Pool) Release(ids ...ID) {
```

Comment:
Will this allocate a slice of IDs under the hood? I'd be in favor of changing it to take one id at a time to minimize heap allocation, since it's only called on two ids at most.